### PR TITLE
Focal L1 Surface Loss: upweight hard surface nodes

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current L1 surface loss treats all surface nodes equally. However, prediction errors concentrate at suction peaks (leading edge), separation points, and trailing edge recovery regions. A **focal-style L1 loss** that upweights high-error nodes should force the model to allocate more capacity to these critical regions, improving p_in and p_tan where suction peak accuracy dominates the error.

This is simpler and more direct than quantile regression (which catastrophically failed at +92-162% by halving gradient magnitude and diluting capacity across 3 output channels). Focal L1 keeps the same single-channel output and the same gradient direction — it only rescales gradient magnitudes.

### Changes to `cfd_tandemfoil/train.py`

1. **Add new flags:**
   ```python
   parser.add_argument('--focal_l1', action='store_true', help='Focal L1 surface loss')
   parser.add_argument('--focal_gamma', type=float, default=1.0, help='Focal L1 gamma exponent')
   ```

2. **Implement focal L1 weighting in the surface loss computation.** Find where the per-node surface L1 loss is computed (likely `F.l1_loss(..., reduction='none')` or similar). Replace the uniform reduction with focal weighting:

   ```python
   if args.focal_l1:
       node_errors = torch.abs(pred_surf - target_surf)
       mean_error = node_errors.detach().mean() + 1e-8
       focal_weight = (node_errors.detach() / mean_error).pow(args.focal_gamma)
       focal_weight = focal_weight.clamp(min=0.5, max=5.0)
       surf_loss = (focal_weight * node_errors).mean()
   ```

3. **Important details:**
   - Use `.detach()` on the error when computing weights — gradient should NOT flow through the weighting function
   - `clamp(min=0.5, max=5.0)` prevents degenerate outlier domination
   - Apply ONLY to the surface loss, not to volume loss or DCT loss
   - Fully compatible with existing hard-node mining

4. **Run 2 seeds:**
   ```bash
   CUDA_VISIBLE_DEVICES=0 python train.py \
     --agent tanjiro --wandb_name "tanjiro/focal-l1-g1.0-s42" --wandb_group focal-l1-surface-loss \
     --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
     --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
     --pressure_first --pressure_deep --residual_prediction --surface_refine \
     --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
     --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
     --focal_l1 --focal_gamma 1.0

   CUDA_VISIBLE_DEVICES=1 python train.py \
     --agent tanjiro --wandb_name "tanjiro/focal-l1-g1.0-s73" --wandb_group focal-l1-surface-loss \
     --seed 73 [same flags as above]
   ```

5. If gamma=1.0 shows promise, try gamma=0.5 in follow-up.

## Baseline
| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| **p_in** | **11.709** | < 11.71 |
| **p_oodc** | **7.544** | < 7.54 |
| **p_tan** | **27.402** | < 27.40 |
| p_re | 6.481 | < 6.48 |

W&B baseline: h6fqcry4 (s42), cuhoscp9 (s73)
Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT tanjiro:

## Results

### Focal L1 Surface Loss (gamma=1.0, 2 seeds)

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline | Δ |
|--------|---------|---------|------------|----------|---|
| **p_in** | 12.26 | 12.62 | **12.44** | 11.709 | **+6.2%** ❌ |
| **p_oodc** | 8.5 | 8.3 | **8.40** | 7.544 | **+11.3%** ❌ |
| **p_tan** | 27.2 | 27.1 | **27.15** | 27.402 | **-0.9%** ✅ |
| p_re | 6.9 | 6.8 | **6.85** | 6.481 | **+5.7%** ❌ |
| val_loss | 0.384 | 0.396 | 0.390 | — | — |

### Details
- **W&B runs:** [a5gi6gnr](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/a5gi6gnr) (s42), [8vq8enj8](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/8vq8enj8) (s73)
- **Group:** focal-l1-surface-loss
- **Epochs:** 160 (180-min wall-clock timeout)
- **Peak VRAM:** 40.1 GB (s42), 40.5 GB (s73)

**Commands used:**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent tanjiro --wandb_name "tanjiro/focal-l1-g1.0-s42" --wandb_group focal-l1-surface-loss \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --focal_l1 --focal_gamma 1.0

CUDA_VISIBLE_DEVICES=1 python train.py \
  --agent tanjiro --wandb_name "tanjiro/focal-l1-g1.0-s73" --wandb_group focal-l1-surface-loss \
  --seed 73 [same flags]
```

### What happened

**The focal L1 loss significantly regresses p_in (+6.2%), p_oodc (+11.3%), and p_re (+5.7%).** Only p_tan shows a minor improvement (-0.9%), which is within noise range.

The hypothesis was that upweighting high-error surface nodes would force the model to allocate more capacity to critical regions (suction peaks, separation points). In practice, this seems to have the opposite effect — by amplifying gradient magnitudes for already-hard nodes, the focal weighting likely:

1. **Destabilized early training** on in-distribution samples, where the model needs to learn smooth pressure distributions first before focusing on hard regions. The focal weighting kicks in from epoch 0, potentially preventing the model from establishing a good baseline representation.

2. **Conflicted with the existing hard-node mining** (active after epoch 30), which already provides a targeted 1.5x weight boost for above-median error nodes on non-tandem samples. The focal L1 on top of hard-node mining creates redundant and potentially excessive upweighting.

3. **Hurt OOD generalization** (p_oodc +11.3%, p_re +5.7%) because focusing on hard nodes in the training distribution doesn't transfer well to out-of-distribution conditions — the "hard" nodes in-distribution are different from the hard nodes OOD.

The modest p_tan improvement may reflect that tandem samples genuinely benefit from more focus on suction peak accuracy, but this is overwhelmed by the regressions elsewhere.

### Suggested follow-ups

- **Delayed focal L1**: Only activate focal weighting after epoch 30 (matching hard-node mining), giving the model time to learn a good base representation first.
- **Lower gamma (0.5)**: The advisor suggested trying gamma=0.5 if 1.0 shows promise. While 1.0 didn't show promise overall, gamma=0.5 might be mild enough to avoid the destabilization.
- **Focal L1 without hard-node mining**: Test focal L1 as a *replacement* for hard-node mining rather than composing them, to avoid the redundant weighting issue.
- **Tandem-only focal**: Apply focal weighting only to tandem samples (where it showed slight benefit), not to single-foil samples.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2346 because ADVISOR: Closing — Focal L1 regressed significantly (p_in +6.2%, p_oodc +11.3%, p_re +5.7%). Only p_tan -0.9% marginal improvement. Root cause: focal weighting from epoch 0 destabilized early learning, and composing with existing hard-node mining created redundant/excessive upweighting. The existing multi-layered difficulty targeting (hard-node mining + PCGrad + tandem boost) already handles hard nodes effectively. Direction exhausted..
